signature/src/policy.rs

```diff
@@ -0,0 +1,893 @@
+    ErrMatchReference,
+    #[strum(to_string = "The policy type name is Unknown.")]
+    ErrUnknownPrType,
+    #[strum(to_string = "The reference match policy type name is Unknown.")]
+    ErrUnknownMatchPolicyType,
+}
+
+// PolicyRequirement is a rule which must be satisfied by the image.
+// It has three types:
+//   `insecureAcceptAnything`: skip signature verification, accept the image unconditionally.
+//   `reject`: reject the image directly.
+//   `signedBy`: there must be at least a signature of the image can be verified by the specific key.
+// The first two are represented by PolicyReqSimple, the last is represented by PolicyReqSignedBy.
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+#[serde(untagged)]
+pub enum PolicyRequirement {
+    SignedBy(PolicyReqSignedBy),
+    Simple(PolicyReqSimple),
+}
+
+impl TryFrom<&str> for PolicyRequirement {
+    type Error = serde_json::Error;
+    fn try_from(json_str: &str) -> Result<Self, Self::Error> {
+        serde_json::from_str::<PolicyRequirement>(json_str)
+    }
+}
+
+impl PolicyRequirement {
+    pub fn is_running_image_allowed(&self, image: &mut image::Image) -> Result<()> {
+        match self {
+            PolicyRequirement::Simple(req) => req.simple_judge_allowed(image),
+            PolicyRequirement::SignedBy(req) => req.signature_judge_allowed(image),
+        }
+    }
+}
+
+// The `insecureAcceptAnything` or `reject` policy requirement.
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+pub struct PolicyReqSimple {
+    r#type: String,
+}
+
+impl PolicyReqSimple {
+    fn simple_judge_allowed(&self, _image: &image::Image) -> Result<()> {
+        if self.r#type == PolicyReqType::TypeReject.to_string() {
+            return Err(anyhow!(r#"The policy is "reject""#));
+        } else if self.r#type == PolicyReqType::TypeAccept.to_string() {
+            Ok(())
+        } else {
+            Err(anyhow!(ErrorInfo::ErrUnknownPrType.to_string()))
+        }
+    }
+}
+
+// The `signedBy` policy requirement:
+// The image must be signed by trusted keys for a specified identity
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+pub struct PolicyReqSignedBy {
+    r#type: String,
+
+    // KeyType specifies what kind of key reference KeyPath/KeyData is.
+    // Acceptable values are "GPGKeys" | "signedByGPGKeys” "X.509Certificates" | "signedByX.509CAs"
+    // FIXME: now only support "GPGKeys", fllowing the [containers/image](https://github.com/containers/image)
+    #[serde(rename = "keyType")]
+    key_type: String,
+
+    // KeyPath is a pathname to a local file containing the trusted key(s).
+    // Exactly one of KeyPath and KeyData must be specified.
+    #[serde(default, rename = "keyPath")]
+    key_path: String,
+    // KeyData contains the trusted key(s), base64-encoded.
+    // Exactly one of KeyPath and KeyData must be specified.
+    #[serde(default, rename = "keyData")]
+    key_data: String,
```

Comment:
This makes me wonder if there is value in creating a type to encapsulate all these related components:

```rust
pub struct Key {
    r#type: String,
    path: String,
    data: String,
}

pub struct PolicyReqSignedBy {
    key: Key,
}

Reply:
The format design of signedBy policy comes from [image simple signing](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md#signedby). I think we'd better keep consistent with their policy format for the time being.